### ISSUE TITLE
Extending a Schema

### ISSUE THREAD

**Issue by VDigitall** [mirumee/ariadne#95]
Hi, it is very useful library, thanks.
Do you plan to implement support extending a schema, [documentation](https://graphql-core-next.readthedocs.io/en/latest/usage/extension.html)?
In my case are default scheme and extensions. I need to extend scheme by some extensions according some rule.

**Comment by rafalp** [comment]
We want to have support for extending schema in Ariadne, but right now we don't have any ideas on how those should look like. Maybe we should add `extensions=` to `make_executable_schema`?

> I need to extend scheme by some extensions according some rule.

Can you tell us more about your use-case? Do you need to extend schema based on application configuration? Or before query is executed, based on authenticated user?

And thanks!

**Comment by VDigitall** [comment]
In my use-case need to extend schema before query is executed, based on authenticated user.
> Maybe we should add extensions= to make_executable_schema?
This is the first thing that came to mind me, too.

**Comment by patrys** [comment]
@VDigitall As a workaround, would it be possible for you to build two schemas and choose the one to execute the query against based on the user? It does not seem like a good idea to rebuild the schema each time a request is made.

**Comment by VDigitall** [comment]
Yes, it is a good idea, but it will be a good approach, when I can store the base schema in one directory and extensions for these schema in the other, and can mix them, instead describing a new full schema.

**Comment by VDigitall** [comment]
For example in my app:
```
my_schema_1 = make_executable_schema(type_defs, extensions=[ext_1, ext_2], resolvers)
my_schema_2 = make_executable_schema(type_defs, extensions=[ext_2, ext_3], resolvers)
my_schema_3 = make_executable_schema(type_defs, extensions=[ext_1, ext_3], resolvers)
```
also in some cases resolvers can be different, too.

**Comment by VDigitall** [comment]
Yes I see docs on modularization.
I mind when I have:
```
user_defs = """
type Product {
  sku: String
  title: String
  keywords: String
  available: Boolean
  price: Float
}"""
```
and now for admin I need to add field `supplier: Supplier`

In your case it would be like:
```
admin_defs = """
type Product {
  sku: String
  title: String
  keywords: String
  available: Boolean
  price: Float
  supplier: Supplier
}"""
```

but for me fit 

```
product_defs = """
type Product {
  sku: String
  title: String
  keywords: String
  available: Boolean
  price: Float
}"""

product_ext_supplier = """
extend type Product {
  supplier: Supplier
}
"""

product_ext_comments = """
extend type Product {
  comments: [Comment]
"""
```
instead.

**Comment by rafalp** [comment]
Maybe we could have an `extend_schema` util so we would take schema, typedefs and resolvers, then return new schema:

```
base_schema = make_executable_schema(type_defs, resolvers)
admin_schema = extend_schema(base_schema, type_defs, resolvers)
assert admin_schema != base_schema
```

**Comment by rmyers** [comment]
@rafalp I added a PR to show a different way to provide this. Rather than making a separate function to call the `make_executable_schema` can add extensions if there are any present in the schema. We can do this by creating an ast document which will contain a list of the definitions in the type defs. Then pull out any that are an extension kind and run extend_schema on that. This is how graphql-tools does it and I find it is very useful when spliting type defs across many files. For example:

`base.graphql`:
```
type People {
      name: String
}

type Query {
      people: [People]
}
```

Now add a completely different type and query in a different file:

`widgets.graphql:`
```
type Widget {
      name: String
      price: Int
}

extend type Query {
       widgets: [Widget]
}
```

This allows you to organize all types and query's in different files and is how apollo-server works. I really want to use Python though :) and would like to use the same pattern in here.

**Comment by Wisdom0063** [comment]
I am also coming from apollo background and I have been wanting to implement this by separating my queries and me able to extend them. It will be a great feature

**Comment by rafalp** [comment]
@Wisdom0063 Then I am happy to let you know this feature has been released in Ariadne 0.4, two months ago ;)

This mechanic is part of `make_executable_schema` logic, so you can just use `extend` keyword in your SDL:

```
type Query {
    someField: Int
}

extend type Query {
    extraField: String
}
```

**Comment by patrys** [comment]
I think Apollo does one more thing where it will synthesize an empty base type when all it gets are `extends`?

**Comment by rafalp** [comment]
No. I've done a quick experiment and they are throwing errors if a type is undefined or empty.

```
Error: Cannot extend type "Query" because it is not defined.
```

**Comment by Wisdom0063** [comment]
> @Wisdom0063 Then I am happy to let you know this feature has been released in Ariadne 0.4, two months ago ;)
> 
> This mechanic is part of `make_executable_schema` logic, so you can just use `extend` keyword in your SDL:
> 
> ```
> type Query {
>     someField: Int
> }
> 
> extend type Query {
>     extraField: String
> }
> ```

Yes. 
I have seen it
Good Job done
